lib/filesystem.nix

```
@@ -148,10 +160,193 @@ in
-    else
-      dir + "/${name}"
+      if type == "directory" then
+        lib.filesystem.listFilesRecursive (dir + "/${name}")
+      else
+        dir + "/${name}"
   ) (builtins.readDir dir));
 
+  /*
+    Transform a directory tree containing Nix packages into an attribute set
+    mapping package names to Nix paths. Most of the time you'll want to use
+    [`packagesFromDirectory`](#function-library-lib.filesystem.packagesFromDirectory)
+    instead, which invokes a `callPackage` function on each of the paths to
+    construct a package set.
+
+    For a directory tree like this:
+
+    ```
+    my-packages
+    ├── a.nix
+    ├── b.nix
+    ├── c
+    │  ├── my-extra-feature.patch
+    │  ├── package.nix
+    │  └── support-definitions.nix
+    └── my-namespace
+       ├── d.nix
+       ├── e.nix
+       └── f
+          └── package.nix
+    ```
+
+    `packageFilesFromDirectory { directory = ./my-packages; }` will return an
+    attribute set like this:
+
+    ```nix
+    {
+      a = ./my-packages/a.nix;
+      b = ./my-packages/b.nix;
+      c = ./my-packages/c/package.nix;
+      my-namespace = {
+        d = ./my-packages/my-namespace/d.nix;
+        e = ./my-packages/my-namespace/e.nix;
+        f = ./my-packages/my-namespace/f/package.nix;
+      };
+    }
+    ```
+
+    In particular:
+    - `.nix` files are returned with the `.nix` extension removed.
+    - Other files are ignored.
+    - If a directory contains a `package.nix`, the directory name is used as
+      the package name.
+    - Otherwise, directories are recursively traversed.
+
+    Example:
+      packageFilesFromDirectory { directory = ./my-packages; }
+      => { ... }
+
+    Type: Path -> AttrSet
+  */
+  packageFilesFromDirectory =
+    # Options
+    {
+      # The directory to read package files from
+      directory
+    , ...
+    }:
+    let
+      supported = basename: type:
+        type == "directory" || hasSuffix ".nix" basename;
+
+      packagesFromSubdirectory = subdirectory:
+        let
+          supportedDirectoryEntries =
+            filterAttrs supported (readDir subdirectory);
+
+          directoryEntryToPackageEntries = basename: type:
+            let
+              path = subdirectory + "/${basename}";
+            in
+            {
+              name = replaceStrings [ ".nix" ] [ "" ] basename;
+
+              value =
+                if type == "regular"
+                then path
+                else if type == "directory"
+                then
+                  let
+                    defaultPackagePath = path + "/package.nix";
+                  in
+                  if pathExists defaultPackagePath
+                  then defaultPackagePath
+                  else packagesFromSubdirectory path
+                else
+                  abort
+                    ''
+                      packagesFromDirectory: Unsupported file type
+
+                      File type: ${type}
+
+                      Path: ${path}
```

Error messages in `lib` should look more like this:
```suggestion
                      lib.filesystem.packagesFromDirectory: Unsupported file type ${type} on path ${toString path}
```

Also very important to use `toString path`. Just using `"${path}"` imports the entire path into the store!